libs/constrained-generators/src/Constrained/Properties.hs

```diff
@@ -69,20 +91,130 @@ prop_conformEmpty ::
-        val <- geval
-        res <- runTerm (singletonEnv var val) tm
-        pure $ QC.property $ conformsToSpec res spec
+prop_univSound :: forall fn. TestableFn fn -> QC.Property
+prop_univSound (TestableFn fn) =
+  QC.label (show fn) $
+    QC.forAllShrinkBlind QC.arbitrary QC.shrink $ \(TestableCtx ctx :: TestableCtx fn as) ->
+      QC.forAllShrinkBlind QC.arbitrary QC.shrink $ \(spec :: Specification fn a) ->
+        QC.counterexample ("\nfn ctx = " ++ showCtxWith fn (TestableCtx ctx)) $
+          QC.counterexample (show $ "\nspec =" <+> pretty spec) $
+            let sspec = simplifySpec (propagateSpecFun fn ctx spec)
+             in QC.counterexample ("\n" ++ show ("propagateSpecFun fn ctx spec =" /> pretty sspec)) $
+                  QC.forAllBlind (strictGen $ genFromSpec @_ @_ @GE sspec) $ \ge ->
+                    fromGEDiscard $ do
+                      a <- ge
+                      let res = uncurryList_ unValue (sem fn) $ fillListCtx ctx $ \HOLE -> Value a
+                      pure $
+                        QC.counterexample ("\ngenerated value: a = " ++ show a) $
+                          QC.counterexample ("\nfn ctx[a] = " ++ show res) $
+                            conformsToSpecProp res spec
+
+prop_gen_sound :: forall fn a. HasSpec fn a => Specification fn a -> QC.Property
+prop_gen_sound spec =
+  let sspec = simplifySpec spec
+   in QC.tabulate "specType spec" [specType spec] $
+        QC.tabulate "specType (simplifySpec spec)" [specType sspec] $
+          QC.counterexample ("\n" ++ show ("simplifySpec spec =" /> pretty sspec)) $
+            QC.forAllBlind (strictGen $ genFromSpec @_ @_ @GE sspec) $ \ge ->
+              fromGEDiscard $ do
+                a <- ge
+                pure $
+                  QC.counterexample ("\ngenerated value: a = " ++ show a) $
+                    conformsToSpecProp a spec
+
+specType :: Specification fn a -> String
+specType TrueSpec {} = "TrueSpec"
+specType SuspendedSpec {} = "SuspendedSpec"
+specType ErrorSpec {} = "ErrorSpec"
+specType MemberSpec {} = "MemberSpec"
+specType TypeSpec {} = "TypeSpec"
+
+showCtxWith ::
+  forall fn as b.
+  (Typeable as, TypeList as, All (HasSpec fn) as, HasSpec fn b) =>
+  fn as b ->
+  TestableCtx fn as ->
+  String
+showCtxWith fn (TestableCtx ctx) = show $ pretty tm
+  where
+    tm :: Term fn b
+    tm =
+      uncurryList (app fn) $
+        fillListCtx (mapListCtxC @(HasSpec fn) (lit @_ @fn . unValue) ctx) (\HOLE -> V $ Var 0)
+
+data TestableFn fn where
+  TestableFn ::
+    ( All (HasSpec fn) as
+    , TypeList as
+    , HasSpec fn b
+    , Typeable as
+    , QC.Arbitrary (Specification fn b)
+    , Typeable (FunTy as b)
+    ) =>
+    fn as b ->
+    TestableFn fn
+
+instance BaseUniverse fn => Show (TestableFn fn) where
+  show (TestableFn (fn :: fn as b)) =
+    show fn ++ " :: " ++ show (typeOf (undefined :: FunTy as b))
```

Comment:
Will the (BaseUniverse fn) restict us in any way?
Are there some Conway types, that have functions?